src/main/java/org/spongepowered/api/service/permission/PermissionService.java

```diff
@@ -29,82 +29,238 @@
 import java.util.Collection;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
 
 
 /**
- * Represents a provider for permissions. This is the interface that a
- * permissions plugin must implement to provide permissions for a user.
+ * Represents a provider of permissions data.
+ *
+ * <p>This is the interface that a permissions plugin must implement and
+ * register to provide permissions for users on the server. Subject related
+ * method calls on Players and Users are redirected to the active permission
+ * service.</p>
+ *
+ * <p>Implementations of PermissionService are expected to be thread-safe.</p>
+ *
+ * <p>Distinctions are made throughout the service between data which is
+ * currently available, and data which may exist, but is not currently
+ * loaded.</p>
+ *
+ * <p>Methods prefixed with "get" will typically only return data if it is
+ * already loaded into memory. Methods prefixed with "load" however will load
+ * data into memory if necessary.</p>
+ *
+ * <p>{@link CompletableFuture}s are used throughout the service in situations
+ * where it is possible that calls may take a considerable time to execute.
+ * {@link CompletableFuture#thenAcceptAsync(Consumer, Executor)} can be used to
+ * run a callback once the method has completed. Alternatively,
+ * {@link CompletableFuture#join()} can be used, but care should be taken to
+ * avoid blocking the server thread.</p>
  */
 public interface PermissionService extends ContextualService<Subject> {
 
+    /**
+     * The standard identifier for the collection which stores users.
+     */
     String SUBJECTS_USER = "user";
```

Comment:
Yeah, it was mentioned.

https://github.com/SpongePowered/SpongeAPI/pull/1524#discussion_r110904194
https://github.com/SpongePowered/SpongeAPI/pull/1524#issuecomment-297090378

> yeah, catalog types do not fit subject collections at all -- they should remain separate things.

I think the general agreement was that they weren't appropriate, but again, I'm sure you know more than I about where they should and should not be used.